lib/ASTGen/Sources/ASTGen/ASTGen.swift

```diff
@@ -112,18 +123,224 @@ struct ASTGenVisitor: SyntaxTransformVisitor {
+  func visit(_ node: EnumCaseParameterClauseSyntax?) -> ASTNode? {
+    guard let node else {
+      return nil
+    }
+
+    return self.visit(node)
+  }
+
+  @inline(__always)
+  func visit(_ node: InheritedTypeListSyntax?) -> BridgedArrayRef {
+    guard let node else {
+      return .init()
+    }
+
+    return self.visit(node)
+  }
+
+  @inline(__always)
+  func visit(_ node: PrecedenceGroupNameListSyntax?) -> BridgedArrayRef {
+    guard let node else {
+      return .init()
+    }
+
+    return self.visit(node)
+  }
+}
+
+extension SyntaxCollection {
+  /// Like ``Sequence.map(_:)``, but returns a `BridgedArrayRef` with a lifetime tied to that of `astgen`.
+  func map<T>(in astgen: ASTGenVisitor, _ transform: (Element) -> T) -> BridgedArrayRef {
+    if self.isEmpty {
+      return .init()
+    }
+
+    let baseAddress = astgen.allocator.allocate(T.self, count: self.count).baseAddress!
+    do {
+      // A loop instead of `initialize(from: self.lazy.map(transform))` because we aren't
+      // doing a great job optimizing the latter.
+      var currentAddress = baseAddress
+      for element in self {
+        currentAddress.initialize(to: transform(element))
+        currentAddress += 1
+      }
+    }
+
+    return .init(data: baseAddress, numElements: SwiftInt(self.count))
+  }
+
+  /// Like ``Sequence.compactMap(_:)``, but returns a `BridgedArrayRef` with a lifetime tied to that of `astgen`.
+  func compactMap<T>(in astgen: ASTGenVisitor, _ transform: (Element) -> T?) -> BridgedArrayRef {
+    if self.isEmpty {
+      return .init()
+    }
+
+    let baseAddress = astgen.allocator.allocate(T.self, count: self.count).baseAddress!
+    do {
+      // A loop instead of `initialize(from: self.lazy.compactMap(transform))` because we aren't
+      // doing a great job optimizing the latter.
+      var currentAddress = baseAddress
+      for element in self {
+        guard let transformed = transform(element) else {
+          continue
+        }
+
+        currentAddress.initialize(to: transformed)
+        currentAddress += 1
+      }
+    }
+
+    return .init(data: baseAddress, numElements: SwiftInt(self.count))
+  }
+}
+
+extension Optional where Wrapped: SyntaxCollection {
+  /// Like ``Sequence.map(_:)``, but returns a `BridgedArrayRef` with a lifetime tied to that of `astgen`.
+  @inline(__always)
+  func map<T>(in astgen: ASTGenVisitor, _ transform: (Wrapped.Element) -> T) -> BridgedArrayRef {
+    switch self {
+    case .none: return .init()
+    case .some(let unwrapped): return unwrapped.map(in: astgen, transform)
+    }
+  }
+}
+
+extension BidirectionalCollection {
+  /// Like ``Sequence.reversed()``, but returns a `BridgedArrayRef` with a lifetime tied to that of `astgen`.
+  func reversed(in astgen: ASTGenVisitor) -> BridgedArrayRef {
+    if self.isEmpty {
+      return .init()
+    }
+
+    let buffer = astgen.allocator.allocate(Element.self, count: self.count)
+    _ = buffer.initialize(from: self.lazy.reversed())
```

Comment:
I've just read your previous discussion with Alex. But `.reversed` is guaranteed _not_ to create new storage.
https://developer.apple.com/documentation/swift/reversedcollection
> * `c.reversed()` does not create new storage

So I don't think we need this method. Please just do
```swift
extension Collection {
  func bridgedArray(in astgen: ASTGenVisitor) -> BridgedArrayRef {
    let buffer = astgen.allocator.allocate(Element.self, count: self.count)
    _ = buffer.initialize(from: self)
    return .init(data: buffer.baseAddress, numElements: SwiftInt(self.count))
  } 
}
```
or
```swift
extension BridgedArrayRef {
  init<C: Collection>(_ collection: C, in astgen: ASTGenVIsitor) {
    let buffer = astgen.allocator.allocate(C.Element.self, count: collection.count)
    _ = buffer.initialize(from: collection)
    self.data = buffer.baseAddress
    self.numElements = SwiftInt(buffer.count)
  }
}
```
And use `.reversed()` at the cal site.

Reply:
> `.reversed()` on `BidirectionalCollection` directly doesn't create an intermediate array

Thank you, I forgot about this!

Reply:
@rintaro I would like to statically enforce calling `bridgedArray` on lazy collections to at least a certain extent so that we don’t have to keep our eyes as open for these intermediate arrays. Do you mind having this instead (with  `map(in:)` removed)?:

```swift
extension LazyCollectionProtocol {
  func bridgedArray(in astgen: ASTGenVisitor) -> BridgedArrayRef {
    ...
  }
}
// 'ReversedCollection' does not conform to 'LazyCollectionProtocol', and cannot here because it only
// conditionally conforms to 'LazySequenceProtocol' in the standard library.
// FIXME: We could make it conform unconditionally
extension ReversedCollection {
  func bridgedArray(in astgen: ASTGenVisitor) -> BridgedArrayRef {
    self.lazy.bridgedArray(in: astgen)
  }
}
```